<body>
	<div id="root"></div>
	<script src="https://unpkg.com/react@16.12.0/umd/react.development.js"></script>
	<script src="https://unpkg.com/react-dom@16.12.0/umd/react-dom.development.js"></script>
	<script src="https://unpkg.com/@babel/standalone@7.8.3/babel.js"></script>

	<!-- this lets us use ReactErrorBoundary -- eliminating a lot of coding and adding more function -->
	<script src="https://unpkg.com/react-error-boundary@1.2.5/dist/umd/react-error-boundary.js"></script>
	<script type="text/babel">
		const ErrorBoundary = ReactErrorBoundary.ErrorBoundary;
		// class ErrorBoundary extends React.Component {
		//   state = {error: null}
		//   static getDerivedStateFromError(error) {
		//     return {error}
		//   }
		//   render() {
		//     const {error} = this.state
		//     if (error) {
		//       return <this.props.FallbackComponent error={error} />
		//     }

		//     return this.props.children
		//   }
		// }

		function ErrorFallback({ error }) {
			return (
				<div>
					<p>Something went wrong:</p>
					<pre>{error.message}</pre>
				</div>
			);
		}

		function Bomb() {
			throw new Error('ðŸ’¥ CABOOM ðŸ’¥');
		}
		// you can choose where to put the error boundary so that some parts of the app can still run if one part breaks -- or you can choose to stop the whole app or parts of the app even if they aren't throwing errors
		function App() {
			const [explode, setExplode] = React.useState(false);
			return (
				<div>
					<div>
						<button onClick={() => setExplode(true)}>ðŸ’£</button>
					</div>
					<div>
						<ErrorBoundary FallbackComponent={ErrorFallback}>
							{explode ? <Bomb /> : 'Push the button Max!'}
						</ErrorBoundary>
					</div>
				</div>
			);
		}
		ReactDOM.render(<App />, document.getElementById('root'));
	</script>
</body>
